Fix eager updateFromV1

masnax opened the pull request:
`updateFromV1` was introduced by https://github.com/canonical/microcluster/pull/94 because our schema update handling logic was fundamentally broken.

The problem is, this update is eager and needs to apply before the cluster has established all systems are ready for the update. This means that `micro* cluster list` will be broken on all running nodes when 1 node runs the update. This is a particular issue for MicroOVN whose focus for this cycle has been hassle-free updates.


This PR attempts to address the problem by maintaining a separate table: `internal_cluster_members_new` so that old cluster members can continue to use the old table. 

Once a cluster member is refreshed to expect the update from `updateFromV1`, it attempts to create this temporary table. All refreshed cluster members will update their expected version in only this table. Once the final cluster member has applied the update, the old `internal_cluster_members` table is replaced with this one.

This means existing cluster members can continue to function, and updated cluster members can use a different table in the meantime.

One caveat here is that this new table can get out of date, since it's only maintained for schema updates. We can get around this by having some sqlite `TRIGGER`s which fire on `INSERT/DELETE` executions, and update the temporary table, if for example a user removes or adds a cluster member in the middle of an update.

masnax commented:
This should fix the error reported in https://github.com/canonical/microceph/issues/367 as not-upgraded cluster members will continue to function with the old table.

tomponline commented:
> However, since this update runs before we have ensured every cluster member is ready for the update, this means that `micro* cluster list` will be broken on all running nodes when 1 node is refreshed with the update. This is a particular issue for MicroOVN whose focus for this cycle has been hassle-free updates.

Why is this?

masnax commented:
> > However, since this update runs before we have ensured every cluster member is ready for the update, this means that `micro* cluster list` will be broken on all running nodes when 1 node is refreshed with the update. This is a particular issue for MicroOVN whose focus for this cycle has been hassle-free updates.
> 
> Why is this?

The schema upgrade process works like in LXD where each member updates its row in the cluster members table with the size of the upgrade list, then compares this value with other rows. If all rows don't match for that column, the node waits for a notification over the internal API. The first cluster member to see all rows match for that column applies the updates. 

Then we compare this value to the latest entry in the `schemas` table to determine which updates to run. Since the value in the old table is an aggregate of internal/external updates, we can't accurately determine which updates we have already run. This is the example I gave with 1+3=4 vs 2+2=4. 

So we need to split the internal/external update count first outside of the main update mechanism. Right now we do this eagerly, so right when the first node starts. 

If we want to run that update after all nodes agree, we need two separate implementations for how to establish that agreement, because after the agreement is established and we run the update, the next time the nodes restart they won't be able to use the old mechanism anymore, since the schema will have changed.

masnax commented:
@tomponline I've modified this PR with a slightly cleaner approach to the same problem. I've updated the PR description as well. I have #154 up as a draft with the old approach that uses a temporary table to compare.

One caveat with this approach (which wasn't present in the temporary table version) is that there is a slight chance of incorrectly determining two cluster members expect the same schema version.

masnax commented:
@roosterfish If you've got any thoughts on this one versus #154, or if you've got a better idea for how to handle this problem, please let me know, thanks :)

masnax commented:
@tomponline @roosterfish I ended up sticking with a new table with insert/delete triggers over adding a view for two reasons:

* If a cluster member is refreshed accidentally to the new version, there is no way to undo the view creation which means the cluster will be forced to update or become read-only. With a temporary table, we can still revert the refresh until all cluster members agree. The only side effect would be a leftover table that is not used by the old implementation anyway.

* A view is not writeable, which means if a cluster member is in an unhealthy state during an update, the cluster will never agree on the update. We would need to add support for something like LXD's `patch.global.sql` to work around this. Using a temporary table means other cluster members can still remove broken nodes, and with the triggers, these changes will be reflected in the temporary table.

roosterfish commented on internal/db/update/update.go:
```diff
@@ -158,12 +158,29 @@ CREATE TABLE internal_cluster_members_new (
+			stmt = "INSERT INTO internal_cluster_members_new SELECT id,name,address,certificate,1,(schema-1),heartbeat,role FROM internal_cluster_members;"
+		}
+
+		stmt = fmt.Sprintf(`%s
```
Should this be in `if count == 0 { } else { "here?" }`. Otherwise the `stmt` statement in case `count == 0` will be overwritten.

masnax replied:
That's okay, we are appending the old `stmt` to the new `stmt`. If `count == 1`, `stmt` will be 2 lines. If `count == 0`, `stmt` will be 3 lines.

roosterfish replied:
Ah yes I see, missed the `fmt.Sprintf("%s ...", stmt)` part :)

roosterfish commented on cluster/cluster_members.go:
```diff
@@ -132,6 +132,22 @@ CREATE TABLE internal_cluster_members_new (
 );
 
 INSERT INTO internal_cluster_members_new SELECT id,name,address,certificate,1,(schema-1),heartbeat,role FROM internal_cluster_members;
+
```
Ah ok, ignore my comment from before, you have added the triggers in a follow up commit. I got confused by the code comment "Also create some triggers ...".

roosterfish reviewed (changes requested):
I understand why having a view won't work as downstreams might want to delete rows from the `internal_cluster_members` table as part of removing some nodes from the cluster.

masnax commented on internal/db/update/schema.go:
```diff
@@ -111,14 +111,25 @@ func (s *SchemaUpdate) Ensure(db *sql.DB) (int, error) {
 		if exists && updateSchemaTable {
-			// updateFromV1 changes the schema table and needs to be run before we calculate the schema version.
-			err := updateFromV1(ctx, tx)
+			versions, err := query.SelectIntegers(ctx, tx, "SELECT COALESCE(MAX(version), 0) FROM schemas ORDER BY version")
```
Ah, good point. Poor copying on my end :)

masnax commented on internal/db/update/schema.go:
```diff
@@ -111,14 +111,25 @@ func (s *SchemaUpdate) Ensure(db *sql.DB) (int, error) {
 		if exists && updateSchemaTable {
-			// updateFromV1 changes the schema table and needs to be run before we calculate the schema version.
-			err := updateFromV1(ctx, tx)
+			versions, err := query.SelectIntegers(ctx, tx, "SELECT COALESCE(MAX(version), 0) FROM schemas ORDER BY version")
```
Oops, yes you're right

masnax commented on cluster/cluster_members.go:
```diff
@@ -86,10 +86,71 @@ func (c InternalClusterMember) ToAPI() (*internalTypes.ClusterMember, error) {
+	}
+
+	// If we have a `type` field, then the default database is valid, so we can just use that.
+	// TODO: Dynamically rename to "core_"
```
Yup, just here as reminders :)

masnax commented on cluster/cluster_members.go:
```diff
@@ -86,10 +86,71 @@ func (c InternalClusterMember) ToAPI() (*internalTypes.ClusterMember, error) {
+  UNIQUE(certificate)
+);
+
+INSERT INTO internal_cluster_members_new SELECT id,name,address,certificate,1,(schema-1),heartbeat,role FROM internal_cluster_members;
```
Right, prior to `updateFromV1`, the internal schema version is 1 (because we only have `updateFromV0`). This if the schema version is `>1`, those are all external updates.

roosterfish replied:
Thanks.

roosterfish reviewed (approved):
LGTM!
